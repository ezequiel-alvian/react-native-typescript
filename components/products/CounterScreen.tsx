import React from 'react'
import { View, Text, StyleSheet, Dimensions } from 'react-native'
import { Button } from '@rneui/themed'
import { IProps } from '../../types/types'

const CounterScreen: React.FC<IProps> = ({count, setCount})=> {

    const handleIcrease = () => setCount(count + 1)
    const handleDecrease = () => {
        if (count === 0) return 
        setCount(count - 1) 
    }

    return(
        <View style={styles.containerButton}>
            <Button 
                onPress={handleIcrease}
                titleStyle={{color:'#000'}} 
                buttonStyle={styles.buttonCustom}
            >+</Button>
            <View>
                <Text>{count}</Text>
            </View>
            <Button
                onPress={handleDecrease}
                titleStyle={{color:'#000'}} 
                buttonStyle={styles.buttonCustom}
            >-</Button>
        </View>
    )
}

const styles = StyleSheet.create({
    containerButton: {
        display:'flex',
        flexDirection:'row',
        justifyContent:'space-between',
        alignItems:'center',
        backgroundColor:'#00FF7F',
        width: (Dimensions.get('window').width -20 ),
    },
    buttonCustom: {
        backgroundColor:'#00FF7F',
    },
})

export default CounterScreen