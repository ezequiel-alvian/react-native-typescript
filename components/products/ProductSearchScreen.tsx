import React, { useState } from 'react'
import { Text, View, StyleSheet, Dimensions } from 'react-native'
import { SearchBar } from '@rneui/themed';



const ProductSearchScreen = () => {
    const [search, setSearch] = useState()

    const updateSearch = (newVal:any) => {
        setSearch(newVal)
    }
    return (
        <View style={styles.container}>
            <SearchBar
                placeholder="Type Here..."
                onChangeText={(newVal)=> updateSearch(newVal)}
                value={search}
                lightTheme={true}
                containerStyle={styles.containerStyle}
                inputContainerStyle={{backgroundColor:'#fff', borderWidth:0}}
            />
            <Text style={styles.containerTitle}>All Items</Text>
        </View>
    )
}

const styles = StyleSheet.create({
    container:{
        padding:5,
    },
    containerStyle:{
        backgroundColor:'transparent', 
        borderBottomColor:'transparent',
        borderTopColor: 'transparent'
    },
    containerTitle:{
        marginLeft:10,
        fontSize:16,
        fontWeight:'bold',
        fontFamily:'Helvetica',
        marginBottom:5,
        marginTop:10
    }
})

export default ProductSearchScreen