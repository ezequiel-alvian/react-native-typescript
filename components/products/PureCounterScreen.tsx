import React from "react"
import { View, Text } from "react-native"
import { Button } from '@rneui/themed'
import { IProps } from '../../types/types'

const PureCounterScreen: React.FC<IProps> = ({ styles, count, setCount }) => {

    const handleIcrease = () => setCount(count + 1)
    const handleDecrease = () => {
        if (count === 0) return 
        setCount(count - 1) 
    }
    
    return (
        <View style={styles.containerButton}>
                        <Button
                onPress={handleDecrease}
                titleStyle={{color:'#000'}} 
                buttonStyle={styles.buttonCustom}
            >-</Button>
            <View>
                <Text>{count}</Text>
            </View>
            <Button 
                onPress={handleIcrease}
                titleStyle={{color:'#000'}} 
                buttonStyle={styles.buttonCustom}
            >+</Button>
        </View>
    )
}
export default PureCounterScreen