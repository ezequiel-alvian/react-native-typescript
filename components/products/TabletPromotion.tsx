import React from 'react'
import { View, Text, StyleSheet } from 'react-native'
import { Dimensions } from "react-native"

const TabletPromotion = ({ promotion } : {promotion:number}) => {
    return(
        <View style={styles.tabletPromotion}>
            <Text style={styles.tabletPromotionText}>{`-${promotion}%`}</Text>
        </View>
    )

}

const styles = StyleSheet.create({
    tabletPromotion: {
        display:'flex',
        justifyContent:'flex-end',
        flexDirection:'column',
        alignItems: 'flex-end',
        width: (Dimensions.get('window').width -40 ),

    },
    tabletPromotionText: {
        padding:5,
        backgroundColor:'#00FF7F',
    }
})

export default TabletPromotion