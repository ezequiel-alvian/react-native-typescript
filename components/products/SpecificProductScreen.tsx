import React,{ useState, useEffect } from 'react'
import { Text, View, StyleSheet } from 'react-native'
import { useSelector } from "react-redux"
import CounterScreen from './CounterScreen'
import { useNavigation } from '@react-navigation/native'
import { ListProduct, Nav, State } from '../../types/types';
import { Button } from '@rneui/themed'
import { v4 as uuidv4 } from 'uuid'
import {addProducts} from '../../reducer/productsReducer'
import { useDispatch } from 'react-redux'
import "react-native-get-random-values"


const SpecificProductScreen = () => {
  const products = useSelector((state: { products:State }) => state.products.todos)
  const { navigate } = useNavigation<Nav>()
  console.log('soy el products',products)
  const [{ product, city, price, location, amount }] = products
  const [count, setCount] = useState(0)
  const dispatch = useDispatch()   

  const onPress = ( products:ListProduct[]) => {
    navigate('Settings')
    const id = uuidv4()
    console.log('soy el products', products)
    dispatch(addProducts(products))
  }

    return(
    <View style={styles.container}>
        <View style={styles.containerImage}>
          <View style={styles.cardImage}></View>
        </View>
        <View style={styles.containerInfo}>
          <View style={styles.displayInfo}>
            <Text>Product</Text>
            <Text>{product}</Text>
          </View>
          <View style={styles.displayInfo}>
            <Text>Ciudad</Text>
            <Text>{city}</Text>
          </View>
          <View style={styles.displayInfo}>
            <Text>Region</Text>
            <Text>{location}</Text>
          </View>
          <View style={styles.displayInfo}>
            <Text>Precio</Text>
            <Text>{price}</Text>
          </View>
          <View style={styles.displayInfo}>
            <Text>Cantidad</Text>
            <Text>{amount}</Text>
          </View>
        </View>
        <CounterScreen
          setCount={setCount}
          count={count}
          />
        <Button
          buttonStyle={{
            backgroundColor:'#00FF7F',
            marginTop:10
          }}
          onPress={()=> onPress(products) }
        >Agregar</Button>
    </View>
    )
  }

  const styles = StyleSheet.create({
    container: {
      flex:1,
      display: 'flex',
      margin:10
    },
    cardImage: {
      width: 320,
      height: 320,
      backgroundColor:'#DAFFCD',
      marginBottom:15
    },
    displayInfo: {
      display:'flex',
      justifyContent:'space-between',
      flexDirection:'row',
      margin:5,
    },
    containerInfo: {
      display:'flex',
      justifyContent:'space-between',
      flexDirection:'column',
      marginTop:15,
      marginBottom:15
    },
    containerImage: {
      display:'flex',
      justifyContent:'center',
      alignItems:'center',
      alignSelf:'center'
    }
  })
export default SpecificProductScreen