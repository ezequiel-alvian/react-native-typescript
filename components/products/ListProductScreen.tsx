import React, { useEffect } from 'react'
import { View, Text, StyleSheet, TouchableHighlight } from 'react-native'
import { ListProduct, Nav } from '../../types/types'
import { ListItem } from '@rneui/themed'
import TabletPromotion from './TabletPromotion'
import { useNavigation } from '@react-navigation/native'
import { useDispatch } from "react-redux"
import { addTasks } from '../../reducer/productsReducer'

const ListProductScreen = ({item}: { item: ListProduct }) => {
    
    const { navigate } = useNavigation<Nav>()
    const dispatch = useDispatch()

    const onPress = ({item}:{item:ListProduct}) => { 
        navigate('SpecificProduct') 
        dispatch(addTasks(item))
    }
    
    return (
        <ListItem style={styles.cardContainer} onPress={()=> onPress({item})}>
            <ListItem.Content key={item.id}>
                    {item.promotion && <TabletPromotion promotion={item.promotion}/>}
                    <View style={styles.cardDisplay}>
                        <View style={styles.cardImage}></View>
                        <View style={styles.containerText}>
                            <ListItem.Title style={styles.title}>{item.product}</ListItem.Title>
                            <ListItem.Subtitle style={styles.subTitle}>{item.city} - {item.location}</ListItem.Subtitle>
                            <Text style={styles.textColor} >${item.price}</Text>
                            <Text style={styles.text}>{item.amount} kg</Text>
                        </View>
                    </View>
            </ListItem.Content>
        </ListItem>
    )
}


const styles = StyleSheet.create({
    cardDisplay:{
        display: 'flex',
        justifyContent:'flex-start',
        flexDirection:'row',
        flex: 1,
    },
    cardContainer: {
        margin: 10,
        borderColor:'#D3D3D3',
        borderBottomWidth:2,
        flex: 1,
        display:'flex',
        flexDirection:'column'
    },
    cardImage: {
        width: 60,
        height: 60,
        backgroundColor:'#00FF7F',
        marginRight:20,
    },
    containerText:{
        marginBottom:5
    },
    title: {
        fontSize:20,
        fontWeight:'bold',
        fontFamily:'Helvetica'
    },
    subTitle: {
        fontSize:14,
        fontWeight:'400',
        color:'#808080',
        fontFamily:'Helvetica',
        marginBottom:8
    },
    text: {
        fontSize:14,
        fontWeight:'400',
        color:'#808080',
        fontFamily:'Helvetica'
    },
    textColor: {
        fontSize:14,
        fontWeight:'600',
        color:'#000',
        fontFamily:'Helvetica'
    }
})


export default ListProductScreen