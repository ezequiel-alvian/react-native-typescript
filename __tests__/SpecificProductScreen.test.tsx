import SpecificProductScreen from '../components/products/SpecificProductScreen';
import products from './data/data.json'
import { fireEvent, render } from "@testing-library/react-native"
import * as redux from 'react-redux';
import useEffect from 'react';

jest.mock('react-redux',() => {
    const actualSelector = jest.requireActual('react-redux')
    return {
        ...actualSelector,
        useSelector:jest.fn()
    }
})

let mockedNavigation= jest.fn(()=> 'SpecificProduct')

jest.mock('@react-navigation/native', () => {
    const actualNav = jest.requireActual(('@react-navigation/native' as any))
return {
    ...actualNav,
    useNavigation: () => ({
        navigate:mockedNavigation
    }),
}
})


describe('<SpecificProductScreen/>',() => {
    beforeEach(()=>{
        const products = {    
            "amount" : 40,
            "available": false,
            "city": "Roca",
            "location": "Roca",
            "name": "Cacho",
            "price": 600,
            "product": "Pera",
            "surname": "Trast",
            "id": "2342i43905493289021893203",
            "promotion": 25
        }
        jest
            .spyOn(redux, 'useSelector')
            .mockReturnValue({
                   products: products // your mock here
                });
    })

    const onPress = jest.fn()
    const setCount= jest.fn()
    const count = 0 
    test('SpecificProductScreen',() => {
        let component = render(<SpecificProductScreen/>)
        expect(component.toJSON()).toMatchSnapshot()
        let button = component.getByText('Agregar')
        fireEvent.press(button)
        let less = component.getByText('-')
        fireEvent.press(less)
        let more = component.getByText('+')
        fireEvent.press(more)
    })
})