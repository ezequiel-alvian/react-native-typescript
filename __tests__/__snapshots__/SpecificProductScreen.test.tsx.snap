// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpecificProductScreen/> SpecificProductScreen 1`] = `
<View
  style={
    {
      "display": "flex",
      "flex": 1,
      "margin": 10,
    }
  }
>
  <View
    style={
      {
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "marginBottom": 15,
        "marginTop": 15,
      }
    }
  >
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 5,
        }
      }
    >
      <Text>
        Product
      </Text>
      <Text />
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 5,
        }
      }
    >
      <Text>
        Ciudad
      </Text>
      <Text />
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 5,
        }
      }
    >
      <Text>
        Region
      </Text>
      <Text />
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 5,
        }
      }
    >
      <Text>
        Precio
      </Text>
      <Text />
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 5,
        }
      }
    >
      <Text>
        Cantidad
      </Text>
      <Text />
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "alignSelf": "center",
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#DAFFCD",
          "height": 320,
          "marginBottom": 15,
          "width": 320,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#00FF7F",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 730,
      }
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": false,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#00FF7F",
              "borderColor": "#2089dc",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "#000",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            +
          </Text>
        </View>
      </View>
    </View>
    <View>
      <Text>
        0
      </Text>
    </View>
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": false,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#00FF7F",
              "borderColor": "#2089dc",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "#000",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            -
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "overflow": "hidden",
        },
        {
          "borderRadius": 2,
        },
        undefined,
        false,
      ]
    }
    testID="RNE_BUTTON_WRAPPER"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": false,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#00FF7F",
            "borderColor": "#2089dc",
            "borderRadius": 2,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 10,
            "padding": 8,
            "paddingHorizontal": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Agregar
        </Text>
      </View>
    </View>
  </View>
</View>
`;
